ssm does not support EC certificates

GGUS ticket [here](https://ggus.eu/index.php?mode=ticket_info&ticket_id=146384).

I am facing the problem that ssmsend only supports RSA certificates, but not EC certificates. Indeed the code only admits RSA certificates.

If I use an EC certificate, I get this error:

```
2020-04-02 13:53:00,696 - ssmsend - INFO - ========================================
2020-04-02 13:53:00,697 - ssmsend - INFO - Starting sending SSM version 2.2.1.
2020-04-02 13:53:00,698 - ssmsend - INFO - Retrieving broker details from ldap://lcg-bdii.cern.ch:2170 ...
2020-04-02 13:53:00,698 - ssm.brokers - DEBUG - Connecting to ldap://lcg-bdii.cern.ch:2170...
2020-04-02 13:53:00,890 - ssmsend - INFO - Found 2 brokers.
2020-04-02 13:53:00,890 - ssmsend - INFO - No server certificate supplied.  Will not encrypt messages.
2020-04-02 13:53:00,945 - ssm.crypto - ERROR - 140577044019016:error:0607907F:digital envelope routines:EVP_PKEY_get1_RSA:expecting an rsa key:p_lib.c:288:

SSM failed to complete successfully.  See log file for details.
2020-04-02 13:53:00,946 - ssmsend - ERROR - SSM failed to complete successfully: Cert and key don't match.
2020-04-02 13:53:00,946 - ssmsend - INFO - SSM has shut down.
2020-04-02 13:53:00,947 - ssmsend - INFO - ========================================
```

This is due to key/cert match verification. If I bypass the verification, I get an error from the servers:

```
2020-04-02 13:56:29,886 - ssmsend - INFO - ========================================
2020-04-02 13:56:29,887 - ssmsend - INFO - Starting sending SSM version 2.2.1.
2020-04-02 13:56:29,887 - ssmsend - INFO - Retrieving broker details from ldap://lcg-bdii.cern.ch:2170 ...
2020-04-02 13:56:29,888 - ssm.brokers - DEBUG - Connecting to ldap://lcg-bdii.cern.ch:2170...
2020-04-02 13:56:30,005 - ssmsend - INFO - Found 2 brokers.
2020-04-02 13:56:30,005 - ssmsend - INFO - No server certificate supplied.  Will not encrypt messages.
2020-04-02 13:56:30,086 - ssm.ssm2 - INFO - Connecting using SSL...
2020-04-02 13:56:30,087 - stomp.py - DEBUG - Attempting connection to host mq.cro-ngi.hr, port 6162
2020-04-02 13:56:30,367 - stomp.py - INFO - Established connection to host mq.cro-ngi.hr, port 6162
2020-04-02 13:56:30,368 - stomp.py - DEBUG - Sent frame: type=CONNECT, headers={}, body=''
2020-04-02 13:56:30,369 - stomp.py - DEBUG - Starting receiver loop
2020-04-02 13:56:30,417 - stomp.py - DEBUG - Received frame: 'ERROR', headers={'message': 'User name [null] or password is invalid. No user for client certificate: ...
2020-04-02 13:56:30,418 - ssm.ssm2 - WARNING - Error message received: java.lang.SecurityException: User name [null] or password is invalid. No user for client certificate: ...
	at org.apache.activemq.security.JaasCertificateAuthenticationBroker.addConnection(JaasCertificateAuthenticationBroker.java:100)
	at org.apache.activemq.security.JaasDualAuthenticationBroker.addConnection(JaasDualAuthenticationBroker.java:100)
	at org.apache.activemq.broker.BrokerFilter.addConnection(BrokerFilter.java:97)
	at org.apache.activemq.broker.MutableBrokerFilter.addConnection(MutableBrokerFilter.java:102)
	at org.apache.activemq.broker.TransportConnection.processAddConnection(TransportConnection.java:809)
	at org.apache.activemq.broker.jmx.ManagedTransportConnection.processAddConnection(ManagedTransportConnection.java:79)
	at org.apache.activemq.command.ConnectionInfo.visit(ConnectionInfo.java:139)
	at org.apache.activemq.broker.TransportConnection.service(TransportConnection.java:334)
	at org.apache.activemq.broker.TransportConnection$1.onCommand(TransportConnection.java:188)
	at org.apache.activemq.transport.MutexTransport.onCommand(MutexTransport.java:45)
	at org.apache.activemq.transport.AbstractInactivityMonitor.onCommand(AbstractInactivityMonitor.java:270)
	at org.apache.activemq.transport.stomp.StompTransportFilter.sendToActiveMQ(StompTransportFilter.java:87)
	at org.apache.activemq.transport.stomp.ProtocolConverter.sendToActiveMQ(ProtocolConverter.java:199)
	at org.apache.activemq.transport.stomp.ProtocolConverter.onStompConnect(ProtocolConverter.java:764)
	at org.apache.activemq.transport.stomp.ProtocolConverter.onStompCommand(ProtocolConverter.java:262)
	at org.apache.activemq.transport.stomp.StompTransportFilter.onCommand(StompTransportFilter.java:75)
	at org.apache.activemq.transport.TransportSupport.doConsume(TransportSupport.java:83)
	at org.apache.activemq.transport.tcp.SslTransport.doConsume(SslTransport.java:102)
	at org.apache.activemq.transport.stomp.StompSslTransportFactory$1$1.doConsume(StompSslTransportFactory.java:70)
	at org.apache.activemq.transport.tcp.TcpTransport.doRun(TcpTransport.java:214)
	at org.apache.activemq.transport.tcp.TcpTransport.run(TcpTransport.java:196)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.security.auth.login.FailedLoginException: No user for client certificate: ...
	at org.apache.activemq.jaas.CertificateLoginModule.login(CertificateLoginModule.java:93)
	at sun.reflect.GeneratedMethodAccessor49.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at javax.security.auth.login.LoginContext.invoke(LoginContext.java:755)
	at javax.security.auth.login.LoginContext.access$000(LoginContext.java:195)
	at javax.security.auth.login.LoginContext$4.run(LoginContext.java:682)
	at javax.security.auth.login.LoginContext$4.run(LoginContext.java:680)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:680)
	at javax.security.auth.login.LoginContext.login(LoginContext.java:587)
	at org.apache.activemq.security.JaasCertificateAuthenticationBroker.addConnection(JaasCertificateAuthenticationBroker.java:86)
	... 21 more

2020-04-02 13:56:30,428 - stomp.py - ERROR - An unhandled exception was encountered in the stomp receiver loop
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/stomp/connect.py", line 741, in __receiver_loop
    self.__notify(frame_type, headers, body)
  File "/usr/lib/python2.6/site-packages/stomp/connect.py", line 718, in __notify
    notify_func(headers, body)
  File "/usr/lib/python2.6/site-packages/ssm/ssm2.py", line 194, in on_error
    raise Ssm2Exception()
Ssm2Exception
2020-04-02 13:56:30,431 - stomp.py - DEBUG - Receiver loop ended
```

--- a/ssm/crypto.py
+++ b/ssm/crypto.py
@@ -63,24 +63,23 @@ def check_cert_key(certpath, keypath):
     if cert == key:
         return False
 
-    p1 = Popen(['openssl', 'x509', '-noout', '-modulus'],
+    p1 = Popen(['openssl', 'x509', '-pubkey', '-noout'],
                stdin=PIPE, stdout=PIPE, stderr=PIPE, universal_newlines=True)
-    modulus1, error = p1.communicate(cert)
+    pubkey1, error = p1.communicate(cert)
 
     if error != '':
         log.error(error)
         return False
 
-    p2 = Popen(['openssl', 'rsa', '-noout', '-modulus'],
+    p2 = Popen(['openssl', 'pkey', '-pubout'],
                stdin=PIPE, stdout=PIPE, stderr=PIPE, universal_newlines=True)
-    modulus2, error = p2.communicate(key)
+    pubkey2, error = p2.communicate(key)
 
     if error != '':
         log.error(error)
         return False
 
-    return modulus1.strip() == modulus2.strip()
-
+    return pubkey1.strip() == pubkey2.strip()
 
 def sign(text, certpath, keypath):
     """Sign the message using the certificate and key in the files specified.

--- a/test/test_crypto.py
+++ b/test/test_crypto.py
@@ -4,6 +4,7 @@ import unittest
 import logging
 import os
 from subprocess import call, Popen, PIPE
+import tempfile
 import quopri
 
 from ssm.crypto import check_cert_key, \
@@ -74,6 +75,19 @@ class TestEncryptUtils(unittest.TestCase):
 
         if check_cert_key(TEST_CERT_FILE, TEST_CERT_FILE):
             self.fail('Accepted certificate as key.')
+
+        # Check incorrect ordering of cert and key path arguments.
+        self.assertFalse(check_cert_key(TEST_KEY_FILE, TEST_KEY_FILE),
+                         'Accepted key as cert.')
+        self.assertFalse(check_cert_key(TEST_KEY_FILE, TEST_CERT_FILE),
+                         'Accepted key and cert wrong way round.')
+
+        # Check behaviour with an invalid cert or key file.
+        with tempfile.NamedTemporaryFile() as tmp:
+            self.assertFalse(check_cert_key(tmp.name, TEST_KEY_FILE),
+                             'Accepted invalid cert file.')
+            self.assertFalse(check_cert_key(TEST_CERT_FILE, tmp.name),
+                             'Accepted invalid key file.')
 
         if not check_cert_key(TEST_CERT_FILE, TEST_KEY_FILE):
             self.fail('Cert and key match but function failed.')

--- a/test/test_crypto.py
+++ b/test/test_crypto.py
@@ -60,37 +60,35 @@ class TestEncryptUtils(unittest.TestCase):
         os.remove(self.ca_certpath)
 
     def test_check_cert_key(self):
-        '''
-        This will print an error log message for the tests that are
-        supposed to fail; you can ignore it.
-        '''
+        """Check that valid cert and key works."""
+        self.assertTrue(check_cert_key(TEST_CERT_FILE, TEST_KEY_FILE),
+                        'Cert and key match but function failed.')
 
-        # One version of the method would have passed this, because of the
-        # way it checked for validity.
-        try:
-            if check_cert_key('hello', 'hello'):
-                self.fail('Accepted non-existent cert and key.')
-        except CryptoException:
-            pass
+    def test_check_cert_key_invalid_paths(self):
+        """Check invalid file paths don't return True."""
+        self.assertFalse(check_cert_key('hello', 'hello'),
+                         'Accepted invalid file paths.')
+        self.assertFalse(check_cert_key(TEST_CERT_FILE, 'k'),
+                         'Accepted invalid key path.')
+        self.assertFalse(check_cert_key('c', TEST_KEY_FILE),
+                         'Accepted invalid cert path.')
 
-        if check_cert_key(TEST_CERT_FILE, TEST_CERT_FILE):
-            self.fail('Accepted certificate as key.')
-
-        # Check incorrect ordering of cert and key path arguments.
+    def test_check_cert_key_arg_order(self):
+        """Check incorrect order of cert and key path args doesn't succeed."""
+        self.assertFalse(check_cert_key(TEST_CERT_FILE, TEST_CERT_FILE),
+                         'Accepted certificate as key.')
         self.assertFalse(check_cert_key(TEST_KEY_FILE, TEST_KEY_FILE),
                          'Accepted key as cert.')
         self.assertFalse(check_cert_key(TEST_KEY_FILE, TEST_CERT_FILE),
                          'Accepted key and cert wrong way round.')
 
-        # Check behaviour with an invalid cert or key file.
+    def test_check_cert_key_invalid_files(self):
+        """Check behaviour with an invalid cert or key file."""
         with tempfile.NamedTemporaryFile() as tmp:
             self.assertFalse(check_cert_key(tmp.name, TEST_KEY_FILE),
                              'Accepted invalid cert file.')
             self.assertFalse(check_cert_key(TEST_CERT_FILE, tmp.name),
                              'Accepted invalid key file.')
-
-        if not check_cert_key(TEST_CERT_FILE, TEST_KEY_FILE):
-            self.fail('Cert and key match but function failed.')
 
     def test_sign(self):
         '''

